### PR TITLE
core/vm, cmd/evm: implement eof validation

### PR DESCRIPTION
**Pull request by holiman**
The bulk of this PR is authored by @lightclient , in the original EOF-work. 
More recently, the code has been picked up and reworked for the new EOF specification, by @MariusVanDerWijden , in https://github.com/ethereum/go-ethereum/pull/29518, and also @shemnon has contributed with fixes. 

They are the main authors to be credited for the code in this PR.

-------------

This PR is an attempt to start eating the elephant one small bite at a time, by selecting only the eof-validation as a standalone piece which can be merged without interfering too much in the core stuff. 

In this PR: 

- [x] Validation of eof containers, lifted from #29518, along with test-vectors from consensus-tests and fuzzing, to ensure that the move did not lose any functionality. 
- [x] Definition of eof opcodes, which is a prerequisite for validation
- [x] Addition of `undefined` to a jumptable entry item. I'm not super-happy with this, but for the moment it seems the least invasive way to do it. A better way might be to go back and allowing nil-items or nil execute-functions to denote "undefined". 
- [x] Gas-calculators. These do not strictly need to be here, I can remove them if we want to reduce the diff even further. 
- [x] benchmarks of eof validation speed 

It would have been nice to break it up further, `core/vm` is growing very very large. However, it's not easy to put `eof` in a submodule. 
- eof validation requires at least the opcodes, and certain opcode-specific info (e.g. stack usage)
- interpreter requires eof to tell it to validate

So what we might do is move the opcodes, and the jumptable definition to a new low-level module. And in `core/vm` keep the jumptable instantiations, so we can pass the actual jumptables to `eof`. Then `eof` doesn't have to import `core/vm`, I think. 

Even with this smaller piece cherry-picked, there's quite a bit of work to do to improve it, better documentation and test coverage. I'll push commits here and rebase quite a bit. 

To do in follow-up PR(s) is to modify the interpreter a bit more in-depth, to operate differently in eof-mode versus non-eof-mode. 

Co-authored-by: lightclient <lightclient@protonmail.com>
Co-authored-by: Marius van der Wijden <m.vanderwijden@live.de>
Co-authored-by: Danno Ferrin <danno.ferrin@shemnon.com>

### CODE REVIEW

File: core/vm/jump_table.go [hunk 1]
```diff
@@ -1059,12 +1073,17 @@ func newFrontierInstructionSet() JumpTable {
 			minStack:   minStack(1, 0),
 			maxStack:   maxStack(1, 0),
 		},
+		INVALID: {
+			execute:  opUndefined,
+			minStack: minStack(0, 0),
+			maxStack: maxStack(0, 0),
+		},
```
Comment (holiman) [inline core/vm/jump_table.go:1080]: This is IMO very wonky, see discussion here: https://github.com/ethereum/go-ethereum/pull/29518/files#r1753395744 .

File: core/vm/validate.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	ErrInvalidMaxStackHeight         = errors.New("invalid max stack height")
+	ErrInvalidCodeTermination        = errors.New("invalid code termination")
+	ErrEOFCreateWithTruncatedSection = errors.New("eofcreate with truncated section")
+	ErrOrphanedSubcontainer          = errors.New("subcontainer not referenced at all")
+	ErrIncompatibleContainerKind     = errors.New("incompatible container kind")
+	ErrStopAndReturnContract         = errors.New("Stop/Return and Returncontract in the same code section")
+	ErrStopInInitCode                = errors.New("initcode contains a RETURN or STOP opcode")
+	ErrTruncatedTopLevelContainer    = errors.New("truncated top level container")
+	ErrUnreachableCode               = errors.New("unreachable code")
+	ErrInvalidNonReturningFlag       = errors.New("invalid non-returning flag, bad RETF")
```
Comment (holiman) [inline core/vm/validate.go:45]: Not all of these methods are mapped in the eofparser error-to-number mapping function. Should they be?

File: core/vm/validate.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+		count++
+		op = OpCode(code[i])
+		if jt[op].undefined {
+			return nil, fmt.Errorf("%w: op %s, pos %d", ErrUndefinedInstruction, op, i)
+		}
+		size := int(immediates[op])
+		if size != 0 && len(code) <= i+size {
+			return nil, fmt.Errorf("%w: op %s, pos %d", ErrTruncatedImmediate, op, i)
+		}
+		switch {
```
Comment (holiman) [inline core/vm/validate.go:92]: use `switch op {`

File: core/vm/memory_table.go [hunk 4]
```diff
@@ -119,3 +120,23 @@ func memoryRevert(stack *Stack) (uint64, bool) {
...
 	return calcMemSize64(stack.Back(0), stack.Back(1))
 }
+
+func memoryExtCall(stack *Stack) (uint64, bool) {
+	x, overflow := calcMemSize64(stack.Back(1), stack.Back(2))
+	if overflow {
+		return 0, true
+	}
+	return x, false
+}
```
Comment (holiman) [inline core/vm/memory_table.go:130]: needlessly wordy

File: core/vm/validate.go [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+	ErrInvalidVersion                = errors.New("invalid version")
+	ErrMissingTypeHeader             = errors.New("missing type header")
+	ErrInvalidTypeSize               = errors.New("invalid type section size")
+	ErrMissingCodeHeader             = errors.New("missing code header")
+	ErrInvalidCodeSize               = errors.New("invalid code size")
+	ErrInvalidContainerSectionSize   = errors.New("invalid container section size")
+	ErrMissingDataHeader             = errors.New("missing data header")
+	ErrMissingTerminator             = errors.New("missing header terminator")
+	ErrTooManyInputs                 = errors.New("invalid type content, too many inputs")
+	ErrTooManyOutputs                = errors.New("invalid type content, too many outputs")
```
Comment (holiman) [inline core/vm/validate.go:56]: Why are these needed, when we already have `ErrInvalidInputs`/`Outputs`?

File: core/vm/validate.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+	ErrInvalidSection0Type           = errors.New("invalid section 0 type, input and output should be zero and non-returning (0x80)")
+	ErrTooLargeMaxStackHeight        = errors.New("invalid type content, max stack height exceeds limit")
+	ErrInvalidContainerSize          = errors.New("invalid container size")
+)
+
+const (
+	NotRefByEither = iota
+	RefByReturnContract
+	RefByEOFCreate
```
Comment (holiman) [inline core/vm/validate.go:65]: unexport (?)

**Comment by holiman** [comment]
Sped it up quite a bit, the slowest by a factor of 10 :) 
```
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/cmd/eofdump
cpu: 12th Gen Intel(R) Core(TM) i7-1270P
                   │  eof.bench.1  │             eof.bench.2             │
                   │    sec/op     │    sec/op     vs base               │
EofParse/test-1-8     1.062µ ± ∞ ¹   1.105µ ± ∞ ¹        ~ (p=0.151 n=5)
EofParse/test-2-8     1.451µ ± ∞ ¹   1.442µ ± ∞ ¹        ~ (p=0.151 n=5)
EofParse/test-3-8     1.770µ ± ∞ ¹   1.533µ ± ∞ ¹  -13.39% (p=0.008 n=5)
EofParse/test-4-8     1.775µ ± ∞ ¹   1.550µ ± ∞ ¹  -12.68% (p=0.008 n=5)
EofParse/test-5-8     2.035µ ± ∞ ¹   1.631µ ± ∞ ¹  -19.85% (p=0.008 n=5)
EofParse/test-6-8     3.484µ ± ∞ ¹   3.039µ ± ∞ ¹  -12.77% (p=0.008 n=5)
EofParse/test-7-8     3.990m ± ∞ ¹   2.896m ± ∞ ¹  -27.43% (p=0.008 n=5)
EofParse/test-8-8     3.868µ ± ∞ ¹   1.225µ ± ∞ ¹  -68.33% (p=0.008 n=5)
EofParse/test-9-8     3.765µ ± ∞ ¹   1.170µ ± ∞ ¹  -68.92% (p=0.008 n=5)
EofParse/test-10-8    4.252µ ± ∞ ¹   1.838µ ± ∞ ¹  -56.77% (p=0.008 n=5)
EofParse/test-11-8    4.727µ ± ∞ ¹   1.829µ ± ∞ ¹  -61.31% (p=0.008 n=5)
EofParse/test-12-8    1.745m ± ∞ ¹   1.106m ± ∞ ¹  -36.63% (p=0.008 n=5)
EofParse/test-13-8    5.778µ ± ∞ ¹   1.836µ ± ∞ ¹  -68.22% (p=0.008 n=5)
EofParse/test-14-8   12.637µ ± ∞ ¹   2.335µ ± ∞ ¹  -81.52% (p=0.008 n=5)
EofParse/test-15-8   12.224µ ± ∞ ¹   2.856µ ± ∞ ¹  -76.64% (p=0.008 n=5)
EofParse/test-16-8    33.30µ ± ∞ ¹   15.61µ ± ∞ ¹  -53.12% (p=0.008 n=5)
EofParse/test-17-8   106.47µ ± ∞ ¹   11.42µ ± ∞ ¹  -89.27% (p=0.008 n=5)
EofParse/test-18-8   337.85µ ± ∞ ¹   27.20µ ± ∞ ¹  -91.95% (p=0.008 n=5)
EofParse/test-19-8    1.368µ ± ∞ ¹   1.365µ ± ∞ ¹        ~ (p=0.889 n=5)
geomean               11.30µ         5.055µ        -55.27%
¹ need >= 6 samples for confidence interval at level 0.95

                   │  eof.bench.1  │               eof.bench.2               │
                   │      B/s      │      B/s        vs base                 │
EofParse/test-1-8    17.96Mi ± ∞ ¹    17.27Mi ± ∞ ¹          ~ (p=0.151 n=5)
EofParse/test-2-8    13.80Mi ± ∞ ¹    13.89Mi ± ∞ ¹          ~ (p=0.151 n=5)
EofParse/test-3-8    12.39Mi ± ∞ ¹    14.31Mi ± ∞ ¹    +15.47% (p=0.008 n=5)
EofParse/test-4-8    12.36Mi ± ∞ ¹    14.15Mi ± ∞ ¹    +14.51% (p=0.008 n=5)
EofParse/test-5-8    11.24Mi ± ∞ ¹    14.04Mi ± ∞ ¹    +24.85% (p=0.008 n=5)
EofParse/test-6-8    21.90Mi ± ∞ ¹    25.11Mi ± ∞ ¹    +14.68% (p=0.008 n=5)
EofParse/test-7-8    11.74Mi ± ∞ ¹    16.17Mi ± ∞ ¹    +37.77% (p=0.008 n=5)
EofParse/test-8-8    15.78Mi ± ∞ ¹    49.84Mi ± ∞ ¹   +215.77% (p=0.008 n=5)
EofParse/test-9-8    16.21Mi ± ∞ ¹    52.18Mi ± ∞ ¹   +221.82% (p=0.008 n=5)
EofParse/test-10-8   7.401Mi ± ∞ ¹   17.118Mi ± ∞ ¹   +131.31% (p=0.008 n=5)
EofParse/test-11-8   6.657Mi ± ∞ ¹   17.204Mi ± ∞ ¹   +158.45% (p=0.008 n=5)
EofParse/test-12-8   13.43Mi ± ∞ ¹    21.20Mi ± ∞ ¹    +57.88% (p=0.008 n=5)
EofParse/test-13-8   5.608Mi ± ∞ ¹   17.662Mi ± ∞ ¹   +214.97% (p=0.008 n=5)
EofParse/test-14-8   4.072Mi ± ∞ ¹   22.058Mi ± ∞ ¹   +441.69% (p=0.008 n=5)
EofParse/test-15-8   11.70Mi ± ∞ ¹    50.09Mi ± ∞ ¹   +328.04% (p=0.008 n=5)
EofParse/test-16-8   15.35Mi ± ∞ ¹    32.74Mi ± ∞ ¹   +113.23% (p=0.008 n=5)
EofParse/test-17-8   6.590Mi ± ∞ ¹   61.455Mi ± ∞ ¹   +832.56% (p=0.008 n=5)
EofParse/test-18-8   2.947Mi ± ∞ ¹   36.602Mi ± ∞ ¹  +1142.07% (p=0.008 n=5)
EofParse/test-19-8   22.31Mi ± ∞ ¹    22.36Mi ± ∞ ¹          ~ (p=0.841 n=5)
geomean              10.65Mi          23.81Mi         +123.64%
¹ need >= 6 samples for confidence interval at level 0.95

                   │   eof.bench.1   │              eof.bench.2               │
                   │      B/op       │     B/op       vs base                 │
EofParse/test-1-8        352.0 ± ∞ ¹     352.0 ± ∞ ¹        ~ (p=1.000 n=5) ²
EofParse/test-2-8        416.0 ± ∞ ¹     416.0 ± ∞ ¹        ~ (p=1.000 n=5) ²
EofParse/test-3-8        464.0 ± ∞ ¹     424.0 ± ∞ ¹   -8.62% (p=0.008 n=5)
EofParse/test-4-8        464.0 ± ∞ ¹     424.0 ± ∞ ¹   -8.62% (p=0.008 n=5)
EofParse/test-5-8        488.0 ± ∞ ¹     440.0 ± ∞ ¹   -9.84% (p=0.008 n=5)
EofParse/test-6-8       1064.0 ± ∞ ¹    1016.0 ± ∞ ¹   -4.51% (p=0.008 n=5)
EofParse/test-7-8     1033.8Ki ± ∞ ¹   910.3Ki ± ∞ ¹  -11.95% (p=0.008 n=5)
EofParse/test-8-8        934.0 ± ∞ ¹     368.0 ± ∞ ¹  -60.60% (p=0.008 n=5)
EofParse/test-9-8        934.0 ± ∞ ¹     368.0 ± ∞ ¹  -60.60% (p=0.008 n=5)
EofParse/test-10-8      1031.0 ± ∞ ¹     472.0 ± ∞ ¹  -54.22% (p=0.008 n=5)
EofParse/test-11-8      1031.0 ± ∞ ¹     472.0 ± ∞ ¹  -54.22% (p=0.008 n=5)
EofParse/test-12-8     442.7Ki ± ∞ ¹   369.0Ki ± ∞ ¹  -16.64% (p=0.008 n=5)
EofParse/test-13-8      1697.0 ± ∞ ¹     472.0 ± ∞ ¹  -72.19% (p=0.008 n=5)
EofParse/test-14-8      3557.0 ± ∞ ¹     568.0 ± ∞ ¹  -84.03% (p=0.008 n=5)
EofParse/test-15-8      3557.0 ± ∞ ¹     984.0 ± ∞ ¹  -72.34% (p=0.008 n=5)
EofParse/test-16-8     12.48Ki ± ∞ ¹   10.69Ki ± ∞ ¹  -14.34% (p=0.008 n=5)
EofParse/test-17-8    29.664Ki ± ∞ ¹   3.297Ki ± ∞ ¹  -88.89% (p=0.008 n=5)
EofParse/test-18-8   107.798Ki ± ∞ ¹   4.789Ki ± ∞ ¹  -95.56% (p=0.008 n=5)
EofParse/test-19-8       408.0 ± ∞ ¹     408.0 ± ∞ ¹        ~ (p=1.000 n=5) ²
geomean                3.065Ki         1.472Ki        -51.96%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal

                   │  eof.bench.1  │              eof.bench.2              │
                   │   allocs/op   │  allocs/op    vs base                 │
EofParse/test-1-8      9.000 ± ∞ ¹    9.000 ± ∞ ¹        ~ (p=1.000 n=5) ²
EofParse/test-2-8      12.00 ± ∞ ¹    14.00 ± ∞ ¹  +16.67% (p=0.008 n=5)
EofParse/test-3-8      16.00 ± ∞ ¹    14.00 ± ∞ ¹  -12.50% (p=0.008 n=5)
EofParse/test-4-8      16.00 ± ∞ ¹    14.00 ± ∞ ¹  -12.50% (p=0.008 n=5)
EofParse/test-5-8      18.00 ± ∞ ¹    14.00 ± ∞ ¹  -22.22% (p=0.008 n=5)
EofParse/test-6-8      35.00 ± ∞ ¹    31.00 ± ∞ ¹  -11.43% (p=0.008 n=5)
EofParse/test-7-8     38.28k ± ∞ ¹   28.30k ± ∞ ¹  -26.08% (p=0.008 n=5)
EofParse/test-8-8      35.00 ± ∞ ¹    11.00 ± ∞ ¹  -68.57% (p=0.008 n=5)
EofParse/test-9-8      35.00 ± ∞ ¹    11.00 ± ∞ ¹  -68.57% (p=0.008 n=5)
EofParse/test-10-8     37.00 ± ∞ ¹    14.00 ± ∞ ¹  -62.16% (p=0.008 n=5)
EofParse/test-11-8     37.00 ± ∞ ¹    14.00 ± ∞ ¹  -62.16% (p=0.008 n=5)
EofParse/test-12-8    18.88k ± ∞ ¹   10.71k ± ∞ ¹  -43.29% (p=0.008 n=5)
EofParse/test-13-8     40.00 ± ∞ ¹    14.00 ± ∞ ¹  -65.00% (p=0.008 n=5)
EofParse/test-14-8     80.00 ± ∞ ¹    14.00 ± ∞ ¹  -82.50% (p=0.008 n=5)
EofParse/test-15-8     80.00 ± ∞ ¹    14.00 ± ∞ ¹  -82.50% (p=0.008 n=5)
EofParse/test-16-8    279.00 ± ∞ ¹    22.00 ± ∞ ¹  -92.11% (p=0.008 n=5)
EofParse/test-17-8    739.00 ± ∞ ¹    11.00 ± ∞ ¹  -98.51% (p=0.008 n=5)
EofParse/test-18-8   2091.00 ± ∞ ¹    11.00 ± ∞ ¹  -99.47% (p=0.008 n=5)
EofParse/test-19-8     12.00 ± ∞ ¹    12.00 ± ∞ ¹        ~ (p=1.000 n=5) ²
geomean                93.51          29.15        -68.82%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal
```

File: core/vm/validate_test.go [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+		if err := container2.UnmarshalBinary(bin, true); err != nil {
+			b.Fatal(err)
+		}
+		if err := container2.ValidateCode(&pragueEOFInstructionSet, true); err != nil {
+			b.Fatal(err)
+		}
+	}
+}
+
+func BenchmarkEOFValidation2(b *testing.B) {
```
Comment (holiman) [inline core/vm/validate_test.go:358]: All the benchmark-function in this file appears to be non-functional, should we nuke them or salvage them, @MariusVanDerWijden ?

example 
```
[user@work vm]$ go test . -run - -bench EOFV
--- FAIL: BenchmarkEOFValidation
    validate_test.go:353: callf into non-returning section: section 1
--- FAIL: BenchmarkEOFValidation2
    validate_test.go:403: callf into non-returning section: section 1
--- FAIL: BenchmarkEOFValidation3
    validate_test.go:452: callf into non-returning section: section 1
```

Comment (MariusVanDerWijden) [reply core/vm/validate_test.go:358]: I will try to salvage

File: core/vm/validate.go (hunk 2 above)
Comment (MariusVanDerWijden) [reply core/vm/validate.go:46]: At some point maybe, for now we've given up on making eofparser match the error numbers

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

File: cmd/eofdump/eofparser.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+	"github.com/urfave/cli/v2"
+)
+
+func init() {
+	jt = vm.NewPragueEOFInstructionSetForTesting()
+}
+
+var (
+	jt       vm.JumpTable
+	errorMap = map[string]int{
```
Comment (holiman) [inline cmd/eofdump/eofparser.go:42]: @MariusVanDerWijden how about nuking everything related to this error mapping?

File: core/vm/validate_linear.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+		default:
+			if imm := int(immediates[op]); imm != 0 {
+				next = append(next, pos+imm)
+			} else {
+				// Simple op, no operand.
+				next = append(next, pos)
+			}
+		}
+		if debugging {
+			fmt.Println(next)
```
Comment (fjl) [inline core/vm/validate_linear.go:172]: 👁️

**Comment by fjl** [comment]
Let's move the new utility commands into cmd/evm. It's better to use an existing command.

**Comment by fjl** [comment]
Would also be nice to remove some complexity in the validation code by moving statements out of `if`

```go
if have, want := currentStackMax, int(metadata[section].outputs)+int(newSection.inputs)-int(newSection.outputs); have != want {
```

should become

```go
wantStack := int(metadata[section].outputs)+int(newSection.inputs)-int(newSection.outputs)
if currentStackMax != wantStack {
```

**Comment by holiman** [comment]
> Let's move the new utility commands into cmd/evm. It's better to use an existing command.

Done

> Would also be nice to remove some complexity in the validation

Done

Also fixed up the `eofdump` command to print nicer output, and read from stdin if `--hex` is not given. Also made the commands use the std logging facilities more, not `fmt.Printf` so much (Note though: for `eofparse` it's part of the spec that it prints out "OK" or "err..." on stdout, so it can be used to compare across clients)

File: cmd/evm/eofparse.go [hunk 10]
```diff
@@ -41,48 +40,40 @@ var (
...
+func eofParseAction(ctx *cli.Context) error {
 	// If `--test` is set, parse and validate the reference test at the provided path.
-	if ctx.IsSet(RefTestFlag.Name) {
+	if ctx.IsSet(refTestFlag.Name) {
 		var (
-			file          = ctx.String(RefTestFlag.Name)
-			executedTests atomic.Int32
-			passedTests   atomic.Int32
+			file          = ctx.String(refTestFlag.Name)
+			executedTests int
```
Comment (holiman) [inline cmd/evm/eofparse.go:48]: filepath.Walk is sequential, there was no need to use atomics here

**Comment by holiman** [comment]
Example invocations of eofdump:

```
$ go run ./cmd/evm/ eofdump  --hex 0xef000101000402000100010400020000800000feaabb 
Header
  - EOFMagic: ef00
  - EOFVersion: 01
  - KindType: 01
  - TypesSize: 0004
  - KindCode: 02
  - KindData: 04
  - DataSize: 0002
  - Number of code sections: 1
    - Code section 0 length: 0001
  - Number of subcontainers: 0
Body
  - Type 0: 00800000
  - Code section 0: 0xfe
  - Data: 0xaabb
```
```
$ yes "0xef000101000402000100010400020000800000feaabb" | head -n2 | go run ./cmd/evm/ eofdump  
Header
  - EOFMagic: ef00
  - EOFVersion: 01
  - KindType: 01
  - TypesSize: 0004
  - KindCode: 02
  - KindData: 04
  - DataSize: 0002
  - Number of code sections: 1
    - Code section 0 length: 0001
  - Number of subcontainers: 0
Body
  - Type 0: 00800000
  - Code section 0: 0xfe
  - Data: 0xaabb

Header
  - EOFMagic: ef00
  - EOFVersion: 01
  - KindType: 01
  - TypesSize: 0004
  - KindCode: 02
  - KindData: 04
  - DataSize: 0002
  - Number of code sections: 1
    - Code section 0 length: 0001
  - Number of subcontainers: 0
Body
  - Type 0: 00800000
  - Code section 0: 0xfe
  - Data: 0xaabb

```

File: core/vm/eof_control_flow.go [hunk 11]
```diff
@@ -83,8 +77,10 @@ func validateControlFlow(code []byte, section int, metadata []*functionMetadata,
 		case JUMPF:
 			arg, _ := parseUint16(code[pos+1:])
 			newSection := metadata[arg]
-			if have, limit := currentStackMax+int(newSection.maxStackHeight)-int(newSection.inputs), int(params.StackLimit); have > limit {
-				return 0, fmt.Errorf("%w: at pos %d", ErrStackOverflow{stackLen: have, limit: limit}, pos)
+			haveLimit := currentStackMax + newSection.maxOuterHeight()
+			wantLimit := int(params.StackLimit)
```
Comment (fjl) [inline core/vm/eof_control_flow.go:81]: I think it's too much like this. It's fine to compare with `int(params.StackLimit)` directly.
BTW the only reason why we need a conversion here is because `StackLimit` is a typed constant. If we removed the type, it would be possible to compare it against any other integer type directly. Might be a useful change to make.

Comment (fjl) [reply core/vm/eof_control_flow.go:81]: i.e.

```
package params

const StackLimit = 1024
```

instead of

```
const StackLimit uint64 = 1024
```

Comment (holiman) [reply core/vm/eof_control_flow.go:81]: For sure, that can easily be done. However, I pushed a different type of change, as an example: https://github.com/ethereum/go-ethereum/pull/30418/commits/363f4ad1edfdb45e34f02ee6aef71cec35b944c2 . If you think this is a good approach, then I can apply the same approach in some more places. Otherwise I'll revert that commit. 


WDYT?

File: core/vm/eof_control_flow.go [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+			currentStackMin += delta
+		case RETF:
+			if currentStackMax != currentStackMin {
+				return 0, fmt.Errorf("%w: max %d, min %d, at pos %d", errInvalidOutputs, currentStackMax, currentStackMin, pos)
+			}
+			have := int(metadata[section].outputs)
+			if have >= maxOutputItems {
+				return 0, fmt.Errorf("%w: at pos %d", errInvalidNonReturningFlag, pos)
+			}
+			if want := currentStackMin; have != want {
```
Comment (fjl) [inline core/vm/eof_control_flow.go:72]: `have != currentStackMin` (and `have` could be renamed to `outputItems` or something)

File: core/vm/eof_control_flow.go (hunk 11 above)
Comment (fjl) [reply core/vm/eof_control_flow.go]: It looks a lot better now!
